lisa/runner.py

```diff
@@ -247,87 +255,76 @@ def _callback_completed(self, results: List[TestResult]) -> None:
-            self._log.debug(f"no more runner found, total {len(remaining_runners)}")
-
-        if self._runners:
-            run_message = notifier.TestRunMessage(
-                status=notifier.TestRunStatus.RUNNING,
-            )
-            notifier.notify(run_message)
 
-            task_manager = TaskManager[List[TestResult]](
-                self._max_concurrency, self._callback_completed
-            )
-            # set the global task manager for cancellation check
-            set_global_task_manager(task_manager)
-            has_more_runner = True
-
-            # run until no more task and all runner are closed
-            while task_manager.wait_worker() or remaining_runners:
-                assert task_manager.has_idle_worker()
-                has_idle_worker = True
-
-                # runners shouldn't mark them done, until all task completed. It
-                # can be checked by test results status or other signals.
-                assert remaining_runners, (
-                    f"no remaining runners, but there are running tasks. "
-                    f"{task_manager._futures}"
-                )
+        run_message = notifier.TestRunMessage(
+            status=notifier.TestRunStatus.RUNNING,
+        )
+        notifier.notify(run_message)
 
-                for runner in remaining_runners:
-                    while not runner.is_done:
-                        # fetch a task and submit
-                        task = runner.fetch_task()
-                        if task:
-                            # the message may be too long to display.
-                            task_message = str(task)
-                            task_message = (
-                                task_message
-                                if len(task_message) < 200
-                                else f"{task_message[:200]}..."
-                            )
-                            self._log.debug(
-                                f"fetched task from {runner.id}: '{task_message}'"
-                            )
-                            task_manager.submit_task(task)
-                        else:
-                            # current runner may not be done, but it doesn't
-                            # have task temporarily. The root runner can start
-                            # tasks from next runner.
-                            break
-                        if not task_manager.has_idle_worker():
-                            has_idle_worker = False
-                            break
+        task_manager = TaskManager[List[TestResult]](
+            self._max_concurrency, self._callback_completed
+        )
+
+        # set the global task manager for cancellation check
+        set_global_task_manager(task_manager)
+        has_more_runner = True
+
+        # run until no idle workers are available and all runner are closed
+        while task_manager.wait_worker() or has_more_runner:
```

Comment:
This will go in the first while loop always, but re-loop only when it has no idle workers left. So it works fine even without adding `or remaining_runners` condition.

Reply:
Fixed

Reply:
The wait_worker will return false, when there is no running task. There is rare condition that, there is no worker is running, but some runners are not declared to done. That may happen when the whole run is near to complete. The `has_more_runner` cannot block this kind of case.